Add df_summary function

kyleam opened the pull request:
Re: #814

---

~~This has a trivial conflict with #816~~

kyleam committed:
Move trace-to-DataFrame functionality to module

kyleam committed:
Use tracetab functions in SQLite backend

kyleam committed:
Add df_summary function

Re: #814

twiecki commented on pymc3/stats.py:
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
+        varnames = trace.varnames
+
+    if stat_funcs is None:
+        stat_funcs = [lambda x: pd.Series(np.mean(x, 0), name='mean_'),
```
I suppose this is done so as not to collide with the `mean()` method? Note that pandas will would still call `mean()` if you do `df.mean()` even with such a column present, but evaluate the column if you did `df.mean`.

kyleam replied:
> I suppose this is done so as not to collide with the `mean()` method?

Yep.

> Note that pandas will would still call `mean()` if you do `df.mean()`
> even with such a column present, but evaluate the column if you did
> `df.mean`.

Is that a feature of a recent pandas?  With the version of pandas I have
at hand (0.16.2), the method call will still work, but trying to select
the column with `.mean` does not (you instead have to use `[mean]`).

twiecki replied:
You're right, I got myself confused. I would still argue that `mean` is the better name. For example, `.describe()` also creates `mean`.

kyleam replied:
Good point about `describe`.   `mean` it is.

twiecki commented on pymc3/stats.py:
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
+        The alpha level for generating posterior intervals. Defaults
+        to 0.05. This is only meaningful when `stat_funcs` is None.
+    batches : int
+      Batch size for calculating standard deviation for
```
Bad indent.

twiecki commented on pymc3/stats.py:
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
+    batches : int
+      Batch size for calculating standard deviation for
+      non-independent samples. Defaults to 100. This is only
+      meaningful if `stat_funcs` when None.
```
when->is?

twiecki commented on pymc3/stats.py:
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
+
+    Returns
+    -------
+    Data frame with summary statistics for each variable
```
Data frame -> pandas.DataFrame

twiecki commented:
See minor comments but otherwise LGTM.

kyleam commented:
Thanks for the comments. Will update.

fonnesbeck commented on pymc3/stats.py:
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
+        correspond to sampling iterations and the second axis
+        represent the flattened variable (e.g., x__0, x__1,...). It
+        should return either
+        1) a Pandas Series instance where each value corresponds to a
```
Function passing via a pandas data structure ("pandas" as a name is not capitalized, FWIW) is not entirely clear to me. Maybe add an example?

kyleam replied:
> Maybe add an example?

Will do.

> "pandas" as a name is not capitalized, FWIW

Thanks.

fonnesbeck commented:
Looks great to me. This is a great idea.

kyleam commented:
Changes since v1:
- Updated for comments.
- Added `extend` argument to `df_summary`.
- Exposed stats.py's `__all__` in top-level namespace.
- Tried to improve `stat_funcs` argument description.

twiecki commented:
:+1: This is great stuff!
